File: leetcode/print-in-order.py
# pylint: disable=method-hidden,not-callable

class Foo:
    def __init__(self):
        pass

    def first(self, printFirst: 'Callable[[], None]') -> None:
        printFirst()
        self.first = True

    def second(self, printSecond: 'Callable[[], None]') -> None:
        if self.first is True:
            printSecond()
            self.second = True
        else:
            for _ in range(100):
                pass
            return self.second(printSecond)

    def third(self, printThird: 'Callable[[], None]') -> None:
        if self.second is True:
            printThird()
        else:
            for _ in range(100):
                pass
            return self.third(printThird)

File: leetcode/print-in-order.c
#include <stdlib.h>
#include <unistd.h>

void printFirst();
void printSecond();
void printThird();

typedef struct {
    volatile int round;
} Foo;

Foo* fooCreate() {
    Foo* obj = (Foo*) malloc(sizeof(Foo));
    obj->round = 1;
    return obj;
}
void incr(Foo* obj) {
    obj->round = obj->round + 1;
}

void sleep_(Foo* obj, int round) {
    while(obj->round != round) {
        sleep(0);
    }
}

void first(Foo* obj) {
    printFirst();
    incr(obj);
}

void second(Foo* obj) {
    sleep_(obj, 2);
    printSecond();
    incr(obj);
}

void third(Foo* obj) {
    sleep_(obj, 3);
    printThird();
}

void fooFree(Foo* obj) {
    free(obj);
}
